allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -271,6 +310,166 @@ def forward(self,  # type: ignore
+        num_question_tokens : ``int``
+        tensor : ``torch.Tensor``
+            Used for copying the constructed list onto the right device.
+
+        Returns
+        -------
+        batch_probabilities : ``torch.autograd.Variable``
+            Has shape ``(batch_size, num_entities+1, num_question_tokens)``.
+            Contains all the probabilities for an entity given a question word.
+        """
+        batch_probabilities = Variable(tensor.data.new(torch.zeros(batch_size,
+                                                                   num_entities+1,
+                                                                   num_question_tokens)))
+        for batch_index, world in enumerate(worlds):
+            cell_type_index, row_type_index = self._get_entity_index_by_type(world)
+
+            # Separate the scores by type, since normalization summed over types.
+            cell_type_scores = torch.index_select(scores[batch_index],
+                                                  dim=0,
+                                                  index=Variable(tensor.data.new(cell_type_index)).long())
+            row_type_scores = torch.index_select(scores[batch_index],
+                                                 dim=0,
+                                                 index=Variable(tensor.data.new(row_type_index)).long())
+
+            probabilities_cell = torch.nn.functional.softmax(torch.transpose(cell_type_scores, 0, 1), dim=0)
+            probabilities_row = torch.nn.functional.softmax(torch.transpose(row_type_scores, 0, 1), dim=0)
+            # (num_entities_per_type, num_question_tokens)
+            probabilities_cell = torch.transpose(probabilities_cell, 0, 1)
+            probabilities_row = torch.transpose(probabilities_row, 0, 1)
+
+            # Adding 1 to all for 0 probability of the null entity.
+            # (num_entities+1, num_question_tokens)
+            probabilities_all = Variable(tensor.data.new(torch.zeros(num_entities+1, num_question_tokens)))
+            for ind, mat in zip(cell_type_index, probabilities_cell):
+                probabilities_all[ind+1] = mat
+            for ind, mat in zip(row_type_index, probabilities_row):
+                probabilities_all[ind+1] = mat
+            batch_probabilities[batch_index] = probabilities_all
+
+        return batch_probabilities
+
+    @staticmethod
+    def _get_entity_index_by_type(world: WikiTablesWorld) -> Tuple[List, List]:
+        """
+        Returns
+        -------
+        cell_type_index: ``List[int]``
+            Indexes for entities starting with fb:cell.
+        row_type_index: ``List[int]``
+            Indexes for entities not starting with fb:cell.
+        """
+        cell_type_index, row_type_index = [], []
+        entities = world.table_graph.entities
+        entity2index = {entity: entity_index for entity_index, entity in enumerate(entities)}
+        for entity in entities:
+            if entity.startswith('fb:cell'):
```

Comment:
This is another place that you have a check for entity type - this means that if (when) we change how we're using entity types, we have to find it in another place to make sure it's consistent.  It'd be better if we can find a way to use the entity type dictionary that we already computed.